security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -231,50 +300,55 @@ func (sf *SecretFetcher) scrtUpdated(oldObj, newObj interface{}) {
 		return
 	}
 
+	t := time.Now()
 	oldScrtName := oscrt.GetName()
-	newScrtName := nscrt.GetName()
-	if oldScrtName != newScrtName {
-		log.Warnf("Failed to update secret: name does not match (%s vs %s).", oldScrtName, newScrtName)
+	resourceName := nscrt.GetName()
+	if oldScrtName != resourceName {
+		log.Warnf("Failed to update secret: name does not match (%s vs %s).", oldScrtName, resourceName)
 		return
 	}
-	oldCert, oldKey, oldRoot := extractCertAndKey(oscrt)
-	newCert, newKey, newRoot := extractCertAndKey(nscrt)
-	if bytes.Equal(oldCert, newCert) && bytes.Equal(oldKey, newKey) && bytes.Equal(oldRoot, newRoot) {
-		log.Debugf("secret %s does not change, skip update", oldScrtName)
+	oldns, oldcans := k8sSecretIntoSecretItem(oscrt, t)
+	ns, cans := k8sSecretIntoSecretItem(nscrt, t)
+
+	nsExists := ns != nil && oldns != nil
+	cansExists := cans != nil && oldcans != nil
+	if (nsExists && bytes.Equal(oldns.CertificateChain, ns.CertificateChain) && bytes.Equal(oldns.PrivateKey, ns.PrivateKey)) &&
+		(cansExists && bytes.Equal(oldcans.RootCert, cans.RootCert)) {
+		log.Debugf("secret %s does not change, skip update", resourceName)
```

Comment:
I think if cans == nil && oldcans == nil, we still return from here.

Reply:
Re-wrote the whole section so it's more readable.